--- ../decomp/Celeste/CS06_BossEnd.cs	2024-05-21 09:02:13.000000000 -0700
+++ celeste/Celeste/CS06_BossEnd.cs	2026-05-04 19:14:34.525229705 -0700
@@ -123,7 +123,7 @@
 		private IEnumerator BadelineCalmDown()
 		{
 			Audio.SetParameter(Audio.CurrentAmbienceEventInstance, "postboss", 0f);
-			badeline.LoopingSfx.Param("end", 1f);
+			// badeline.LoopingSfx.Param("end", 1f);
 			yield return 0.5f;
 			badeline.Sprite.Play("scaredTransition");
 			Input.Rumble(RumbleStrength.Light, RumbleLength.Long);