--- ../decomp/Celeste/Level.cs	2024-05-21 09:02:13.000000000 -0700
+++ celeste/Celeste/Level.cs	2026-05-04 20:58:16.061907255 -0700
@@ -2159,6 +2159,7 @@
 			{
 				menu.Add(new TextMenu.Header(Dialog.Clean("menu_pause_title")));
 			}
+			menu.Add(new TextMenu.SubHeader("LEGACY AOT-ST BUILD"));
 			menu.Add(new TextMenu.Button(Dialog.Clean("menu_pause_resume")).Pressed(delegate
 			{
 				menu.OnCancel();