--- ../decomp/Celeste/OuiMainMenu.cs	2024-05-23 17:50:24.000000000 -0700
+++ celeste/Celeste/OuiMainMenu.cs	2026-05-04 21:00:01.293478875 -0700
@@ -241,6 +241,7 @@
 					button.Render();
 				}
 			}
+			ActiveFont.DrawOutline("LEGACY AOT-ST BUILD.\nLegal port: celeste.r58playz.dev", new Vector2(80f, 1010f), new Vector2(0f, 1f), Vector2.One * 0.5f, Color.White, 2f, Color.Black);
 		}
 
 		private void OnDebug()